File: python/examples/providers/openai_compatible.py
import os

from mirascope import llm

# Route grok/ models through the OpenAI provider
# using xAI's OpenAI-compatible endpoint
llm.register_provider(
    "openai",
    scope="grok/",
    base_url="https://api.x.ai/v1",
    api_key=os.environ["XAI_API_KEY"],
)


@llm.call("grok/grok-4-latest")
def recommend_book(genre: str):
    return f"Recommend a {genre} book."


response = recommend_book("fantasy")
print(response.text())
print(response.provider_id)  # "openai" - routed through OpenAI provider

File: typescript/examples/providers/openai-compatible.ts
/**
 * OpenAI-compatible provider example.
 *
 * Demonstrates routing models through the OpenAI provider using
 * an OpenAI-compatible endpoint (like xAI, Together, etc.).
 *
 * Run with: bun run example examples/providers/openai-compatible.ts
 */
import { llm } from "mirascope";

// Route grok/ models through the OpenAI provider
// using xAI's OpenAI-compatible endpoint
llm.registerProvider("openai", {
  scope: "grok/",
  baseURL: "https://api.x.ai/v1",
  apiKey: process.env.XAI_API_KEY,
});

const recommendBook = llm.defineCall<{ genre: string }>({
  model: "grok/grok-4-latest",
  template: ({ genre }) => `Recommend a ${genre} book.`,
});

const response = await recommendBook({ genre: "fantasy" });
console.log(response.text());
